### DISCUSSION TITLE
Custom trained recognition model giving Null output

### DISCUSSION THREAD

**Discussion by ghost** [mindee/doctr#1084] [General]
@felixdittrich92 
Hello, I am trying to train the recognition model (base model - crnn_vgg16) using the script provided in https://github.com/mindee/doctr/tree/main/references/recognition, I am able to train the model on a custom dataset (English) and save the weights. I am also able to load the weights and predict on a sample image but I am encountering " " (Null String) as my predicted value for all the boxes. Just to get started I have only used around 4000 isolated word sample images and 10 epochs. I want to know whether the prediction is due to the less number of samples and less training or is it due to any specific mistake which attributes to these null string predictions.

Thank You

**Comment by felixdittrich92** [comment]
Hi @arunpurohit3799 👋 ,
to clarify your questions i need some more information :)

Could you please share the full command + output while training (TextMatch and loss) and inference code you used ? 

Normally you could use the base pretrained models which are trained on french vocab (does fully include english vocab). Or is it not a good fit with your use case ? :)

Training from scratch (without --pretrained) needs much much more data (pretrained models are trained on ~10M samples) / fine tuning (by passing --pretrained to the command) depends a bit on your use case. In an easy case like printed regular text it should be possible with much less data. 

As you can see it would be useful for us if you can share 1 or 2 example images to get some better understanding of your use case (if possible)

**Reply by ghost** [reply]
Thank you @felixdittrich92 

I was able to fine-tune the model on the base pretrained model. 

Earlier, for pretraining from scratch I used the command `python train_tensorflow.py crnn_vgg16_bn --train_path <path_to_train_data> --val_path <path_to_val_data> `

Here is the output while training: 

> Namespace(amp=False, arch='crnn_vgg16_bn', batch_size=4, epochs=10, find_lr=False, font='FreeMono.ttf,FreeSans.ttf,FreeSerif.ttf', input_size=32, lr=0.001, max_chars=12, min_chars=1, name=None, pretrained=False, push_to_hub=False, resume=None, show_samples=False, test_only=False, train_path='/content/drive/MyDrive/OCR_Finetuning/train', train_samples=1000, val_path='/content/drive/MyDrive/OCR_Finetuning/test', val_samples=20, vocab='french', wb=False, workers=None)
Validation set loaded in 0.6581s (31 samples in 8 batches)
Downloading https://doctr-static.mindee.com/models?id=v0.4.1/vgg16_bn_r-c5836cea.zip&src=0 to /root/.cache/doctr/models/vgg16_bn_r-c5836cea.zip
54929408it [00:03, 14186239.00it/s]                 
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
DEBUG:tensorflow:Layer lstm_1 will use cuDNN kernels when running on GPU.
Train set loaded in 3.188s (423 samples in 105 batches)
  0% 0/105 [00:00<?, ?it/s]WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/ctc_ops.py:1443: alias_inplace_add (from tensorflow.python.ops.inplace_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Prefer tf.tensor_scatter_nd_add, which offers the same functionality with well-defined read-write semantics.
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/ctc_ops.py:1443: alias_inplace_add (from tensorflow.python.ops.inplace_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Prefer tf.tensor_scatter_nd_add, which offers the same functionality with well-defined read-write semantics.
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/ctc_ops.py:1426: alias_inplace_update (from tensorflow.python.ops.inplace_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Prefer tf.tensor_scatter_nd_update, which offers the same functionality with well-defined read-write semantics.
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/ctc_ops.py:1426: alias_inplace_update (from tensorflow.python.ops.inplace_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Prefer tf.tensor_scatter_nd_update, which offers the same functionality with well-defined read-write semantics.
100% 105/105 [02:00<00:00,  1.15s/it]
Validation loss decreased inf --> 18.8384: saving state...
100% 105/105 [00:39<00:00,  2.68it/s]
Validation loss decreased 18.8384 --> 18.4063: saving state...
100% 105/105 [00:38<00:00,  2.73it/s]
Validation loss decreased 18.4063 --> 18.3419: saving state...
100% 105/105 [00:39<00:00,  2.69it/s]
Validation loss decreased 18.3419 --> 18.3302: saving state...
100% 105/105 [00:38<00:00,  2.75it/s]
Validation loss decreased 18.3302 --> 18.3193: saving state...
100% 105/105 [00:38<00:00,  2.73it/s]
100% 105/105 [00:38<00:00,  2.76it/s]
Validation loss decreased 18.3193 --> 18.3185: saving state...
100% 105/105 [00:39<00:00,  2.68it/s]
Validation loss decreased 18.3185 --> 18.3183: saving state...
100% 105/105 [00:38<00:00,  2.75it/s]
Validation loss decreased 18.3183 --> 18.3182: saving state...
100% 105/105 [00:38<00:00,  2.75it/s]
Validation loss decreased 18.3182 --> 18.3176: saving state...

For Inference, I used the following code : 

`image = DocumentFile.from_images('/content/test_doc.jpeg')`
`reco_model = from_hub("path_to_the_repo")`
`pred_model = ocr_predictor(det_arch='db_resnet50', reco_arch=reco_model, pretrained=True)`
`result = pred_model(image)`

Attaching a sample image which I used : 

![test_doc](https://user-images.githubusercontent.com/84079786/194479782-3e37f91c-3796-49de-b747-4dba54a904a5.jpeg)

Result for the above image : 

>Document(
  (pages): [Page(
    dimensions=(332, 410)
    (blocks): [
      Block(
        (lines): [Line(
          (words): [
            Word(value='', confidence=3.1e-06),
            Word(value='', confidence=0.18),
          ]
        )]
        (artefacts): []
      ),
      Block(
        (lines): [Line(
          (words): [Word(value='', confidence=0.0094)]
        )]
        (artefacts): []
      ),
      Block(
        (lines): [Line(
          (words): [
            Word(value='', confidence=0.00043),
            Word(value='', confidence=0.17),
          ]
        )]
        (artefacts): []
      ),
      Block(
        (lines): [Line(
          (words): [Word(value='', confidence=0.18)]
        )]
        (artefacts): []
      ),
    ]
  )]
)

As you can see in the result, the value is '' for all the boxes. Since I trained with less amount of data, I thought maybe I could get some random characters or some characters as predictions but here it is completely empty. Please let me know your views on this

Thank you

**Reply by felixdittrich92** [reply]
Hi @arunpurohit3799 :wave:, 
i think the main issue is the image size (for detection it will be resized to (1024, 1024) which leads to wrong/bad detection results for small sized images which afterwards effect also the recognition part.

A workaround for the moment is padding: (tested with your provided image detections and predictions are correct - without any further training)

```
from doctr.io import DocumentFile
from doctr.io.image import tensor_from_numpy
from doctr.models import ocr_predictor
from PIL import Image
import numpy as np

# pad image with black border to (1024, 1024)
img = Image.open('/home/felix/Desktop/handwritten_test.jpeg').convert('RGB')
old_size = img.size
new_size = (1024, 1024)
new_im = Image.new("RGB", new_size)
box = tuple((n - o) // 2 for n, o in zip(new_size, old_size))
new_im.paste(img, box)
# convert to numpy array and add batch dimension
doc = tensor_from_numpy(npy_img=np.array(new_im))
doc = np.expand_dims(doc, axis=0)
# pred
predictor = ocr_predictor('db_resnet50', 'crnn_mobilenet_v3_small', pretrained=True)
result = predictor(doc)
result.show(doc)
```

![Screenshot from 2022-10-07 08-54-39](https://user-images.githubusercontent.com/23098070/194486301-d3495db6-acc3-4904-8953-dffaf6eb3b00.png)

```
Document(
  (pages): [Page(
    dimensions=(1024, 1024)
    (blocks): [
      Block(
        (lines): [Line(
          (words): [Word(value='Because', confidence=0.71)]
        )]
        (artefacts): []
      ),
      Block(
        (lines): [Line(
          (words): [
            Word(value='be', confidence=0.99),
            Word(value='doing', confidence=0.92),
          ]
        )]
        (artefacts): []
      ),
    ]
  )]
)
```

A side note: We do not provide a well working pipeline for handwritten recognition currently (but it's on long term track) :) 

About the training: 
with a loss of 18 I am not surprised that the model is unable to predict anything :sweat_smile: 
However, there could be many reasons why it learns so slowly. 
There are two most likely reasons: to less data and/or something is wrong with your dataset (but I would have to validate this to be able to judge it more precisely).
Maybe the given workaround will help you.
Let me know :hugs:

**Reply by khawar-islam** [reply]
Hi @felixdittrich92 I have added padding in image size is `(1024x1024) `and now i am testing on your above code and getting tensor mismtach error.

**Image**

![rsz_imageedit_7_6413275781](https://user-images.githubusercontent.com/11488932/195297545-d5fd07d8-93a2-4e35-aaba-02578fe9fca7.png)


**Code**
```
import os
os.environ['USE_TORCH'] = '1'
import torch
from doctr.models import ocr_predictor, master
from collections import OrderedDict
from doctr.datasets import VOCABS
from doctr.io import DocumentFile
from doctr.io.image import tensor_from_numpy
from doctr.models import ocr_predictor
from PIL import Image
import numpy as np

# pad image with black border to (1024, 1024)
img = Image.open('/home/cvpr/Downloads/rsz_imageedit_7_6413275781.png').convert('RGB')
old_size = img.size
new_size = (1024, 1024)
new_im = Image.new("RGB", new_size)
box = tuple((n - o) // 2 for n, o in zip(new_size, old_size))
new_im.paste(img, box)
# convert to numpy array and add batch dimension
doc = tensor_from_numpy(npy_img=np.array(new_im))
doc = np.expand_dims(doc, axis=0)
# pred
model = master(pretrained=False, vocab=OrderedDict.fromkeys(VOCABS['korean']))
model.load_state_dict(torch.load("/media/cvpr/CM_22/doctr/references/recognition/master_20221011-161521.pt", map_location="cpu"))
predictor = ocr_predictor(det_arch="db_resnet50", reco_arch=model)
model = predictor(doc)
result.show(doc)
```

**Traceback**

```
Traceback (most recent call last):
  File "/media/cvpr/CM_22/doctr/demo_short.py", line 58, in <module>
    model = predictor(doc)
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/media/cvpr/CM_22/doctr/doctr/models/predictor/pytorch.py", line 96, in forward
    loc_preds = self.det_predictor(pages, **kwargs)
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/media/cvpr/CM_22/doctr/doctr/models/detection/predictor/pytorch.py", line 46, in forward
    processed_batches = self.pre_processor(pages)
  File "/media/cvpr/CM_22/doctr/doctr/models/preprocessor/pytorch.py", line 121, in __call__
    batches = list(multithread_exec(self.normalize, batches))
  File "/media/cvpr/CM_22/doctr/doctr/utils/multithreading.py", line 47, in multithread_exec
    results = map(lambda x: x, tp.map(func, seq))
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/multiprocessing/pool.py", line 771, in get
    raise self._value
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torchvision/transforms/transforms.py", line 269, in forward
    return F.normalize(tensor, self.mean, self.std, self.inplace)
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torchvision/transforms/functional.py", line 360, in normalize
    return F_t.normalize(tensor, mean=mean, std=std, inplace=inplace)
  File "/home/cvpr/anaconda3/envs/pytesseract/lib/python3.9/site-packages/torchvision/transforms/functional_tensor.py", line 959, in normalize
    tensor.sub_(mean).div_(std)
RuntimeError: The size of tensor a (1024) must match the size of tensor b (3) at non-singleton dimension 1

```

**Reply by felixdittrich92** [reply]
@khawar-islam please set USE_TORCH=1 before your script like: USE_TORCH=1 python3 /home/user/script.py  

it definitly improves the detection part a bit:
![Screenshot from 2022-10-12 11-35-42](https://user-images.githubusercontent.com/23098070/195307699-16d90b5d-5164-4e0c-9062-5f2d95aa1e62.png)

**Reply by felixdittrich92** [reply]
I have already opened a ticket to solve this: #1095

**Reply by khawar-islam** [reply]
Hi @felixdittrich92 I have added your suggestion but it didn't give any performance improvement. Almost the results are same.
`USE_TORCH=1 python3 demo.py`

**Reply by felixdittrich92** [reply]
@khawar-islam yeah we have to implement this change on our side first.
Unfortunately i have not the time to do this currently. Maybe @odulcy-mindee or @aminemindee have time to solve it ? (ref #1095)

**Reply by khawar-islam** [reply]
@felixdittrich92 no problem, take your time
